src/main/java/net/minecraftforge/common/BiomeDictionary.java

```diff
@@ -380,7 +380,7 @@ static void ensureHasTypes(Biome biome)
         if (!hasAnyType(biome))
         {
             makeBestGuess(biome);
-            FMLLog.log.warn("No types have been added to Biome {}, types have been assigned on a best-effort guess: {}", biome.getRegistryName(), getTypes(biome));
+            FMLLog.log.warn("No types have been added to Biome {}, types have been assigned on a best-effort guess: {}", biome.getRegistryName(), !getBiomeInfo(biome).types.isEmpty() ? getBiomeInfo(biome).types : "could not guess types");
```

Comment:
This change preserves the method call of `#getTypes` in the bytecode without ever invoking it.
```suggestion
            FMLLog.log.warn("No types have been added to Biome {}, types have been assigned on a best-effort guess: {}", biome.getRegistryName(), !getBiomeInfo(biome).types.isEmpty() ? getBiomeInfo(biome).types : "could not guess types");
            if (Boolean.valueOf(false))
                getTypes(biome);
```

Reply:
That's the point.

> This change preserves the method call of `#getTypes` in the bytecode without ever invoking it.

There may be a handful of bug-fix mods for 1.12.2 Forge that fix this very problem. If we don't include this line, it could cause modpacks to crash when using those mods.

Reply:
ah yes that makes sense

Reply:
let me commit it

Reply:
I think it committed it wrong

Reply:
it is fine now :D